net, pcp: handle multi-part responses and filter for default route while querying default gateway

willcl-ark opened the pull request:
...for default route in pcp pinholing.

Currently we only make a single recv call, which trucates results from large routing tables, or in the case the kernel may split the message into multiple responses (which may happen with `NLM_F_DUMP`).

We also do not filter on the default route. For IPv6, this led to selecting the first route with an `RTA_GATEWAY` attribute, often a non-default route instead of the actual default. This caused PCP port mapping failures because the wrong gateway was used.

Fix both issues by adding multi-part handling of responses and filter for the default route.

Limit responses to ~ 1MB to prevent any router-based DoS.

willcl-ark commented:
cc @laanwj

This patch maintains the FreeBSD-style querying/filtering we have currently, but but increases the size of the response processed to a maximum of 1MB.

laanwj commented:
Concept ACK

willcl-ark commented on src/common/netif.cpp:
```diff
@@ -82,42 +86,64 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
-                std::memcpy(&gw, RTA_DATA(rta_gateway), sizeof(gw));
-                return CNetAddr(gw, scope_id);
+        for (nlmsghdr* hdr = (nlmsghdr*)response; NLMSG_OK(hdr, recv_result); hdr = NLMSG_NEXT(hdr, recv_result)) {
+            if (hdr->nlmsg_type == NLMSG_DONE) {
```
I don't think it is, but like you am unsure about what the guarantees of the protocol are here.

I kind of reverse engineered this looking at miniupnpc and netlink sources, along with an `strace` of `ip route show`, which is where the repeated `recv` calls jumped out to me as a difference between our code and other tools.

This approach simply relies on receiving an `NLMSG_DONE` to signal the end of the response and break. This should handle both single and multi-part messages. Here's a snipped sample of `strace -e filter=recvmsg ip route show` on my system:

<details>
<summary>Details</summary>

```
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=NULL, iov_len=0}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_TRUNC}, MSG_PEEK|MSG_TRUNC) = 2932
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=[[{nlmsg_len=1444, nlmsg_type=RTM_NEWLINK, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1743240135, nlmsg_pid=3248093}, <snip> 0) = 2932
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=NULL, iov_len=0}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_TRUNC}, MSG_PEEK|MSG_TRUNC) = 3384
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=[[{nlmsg_len=1504, nlmsg_type=RTM_NEWLINK, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1743240135, nlmsg_pid=3248093}, <snip> 0) = 3384
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=NULL, iov_len=0}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_TRUNC}, MSG_PEEK|MSG_TRUNC) = 5196
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=[[{nlmsg_len=1488, nlmsg_type=RTM_NEWLINK, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1743240135, nlmsg_pid=3248093}, <snip> 0) = 5196
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=NULL, iov_len=0}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_TRUNC}, MSG_PEEK|MSG_TRUNC) = 20
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=[{nlmsg_len=20, nlmsg_type=NLMSG_DONE, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1743240135, nlmsg_pid=3248093}, 0], iov_len=32768}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 20
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=NULL, iov_len=0}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_TRUNC}, MSG_PEEK|MSG_TRUNC) = 424
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=[[{nlmsg_len=76, nlmsg_type=RTM_NEWADDR, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1743240136, nlmsg_pid=3248093}, <snip> 0) = 424
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=NULL, iov_len=0}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_TRUNC}, MSG_PEEK|MSG_TRUNC) = 600
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=[[{nlmsg_len=72, nlmsg_type=RTM_NEWADDR, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1743240136, nlmsg_pid=3248093}, <snip> 0) = 600
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=NULL, iov_len=0}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_TRUNC}, MSG_PEEK|MSG_TRUNC) = 20
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=[{nlmsg_len=20, nlmsg_type=NLMSG_DONE, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1743240136, nlmsg_pid=3248093}, 0], iov_len=32768}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 20
```

</details>

`NLM_F_MULTI` is indeed set, even in the `NLMSG_DONE` packet.

One other thing I did read about, but not implement, is checking of the sequence number on each message to ensure it was meant for our request. But as we only make a single request i thought this should be OK to omit.

laanwj replied:
Thanks!

The receive flow [here](https://www.infradead.org/~tgr/libnl/doc/core.html#core_recv) seems to indicate:
- Process packet
- If `NLM_F_MULTI` is set and the packet is not `NLMSG_DONE`, repeat

What i'm mostly worried about is that the current code will hang if `NLMSG_DONE` never comes, which seems to be the case for non-multi responses, which have one data packet.

But it may be that the `NETLINK_ROUTE` response to `NLM_F_DUMP` is always multi-packet. That empirically seems to be the case even for tiny routing tables.

Looking at the `ip` source is a good idea. Also we need to verify this with FreeBSD.

>  But as we only make a single request i thought this should be OK to omit.

Agree, going that far in checking seems unnecessary. i don't think we need super defensive coding as netlink is a local protocol with the kernel.

willcl-ark replied:
Thanks, that infradead page is very handy!

I made a few changes in 6c694c212f8d898dac9cc1c5637381452c91e79b based on your thoughts (and the protocol page):

- set socket to non-blocking mode to avoid hanging if the kernel doesn't send a response
- use a vector to collect all data from multi-part responses
- exit when `recv()` returns 0 (this should handle single-part messages, AFAICT)

I think relying on receiving no more data from `recv()` to break the receive loop should be as (or perhaps more) robust than checking for the `NLM_F_MULTI` flag and exiting after first receive if it's _not_ set, but curious what you think here?

If it would help, I'd be happy to break this into a few smaller commits, as I'm kinda feeling this change contains a few different changes in one in some ways now...

laanwj replied:
> I think relying on receiving no more data from recv() to break the receive loop should be as (or perhaps more) robust than checking for the NLM_F_MULTI flag and exiting after first receive if it's not set, but curious what you think here?

Maybe-is it safe to assume that netlink will never block?

We don't want to end up in the same situation as before where we miss data. But due to say, a threading race condition.

i think the safest thing here is to mimic as closely as possible `ip`'s behavior, as it is the only tool these kind of interfaces tend to be written towards.

laanwj replied:
> use a vector to collect all data from multi-part responses

i'm not sure i see the motivation here. Parsing the packets as seperate units is just as valid, avoids dynamic allocation, and is simpler.

Sjors replied:
> checking of the sequence number on each message to ensure it was meant for our request

This seems like a good idea to at least do in debug builds.

It seems like a good precaution to check for the presence of `NLM_F_MULTI` and don't wait for `NLMSG_DONE` if it isn't. At least from my naive reading of https://man7.org/linux/man-pages/man7/netlink.7.html it seems `NLMSG_DONE` is only used for multipart messages.

Splitting into multiple commits would be useful, e.g. one commit that switches to non-blocking mode.

Sjors commented on src/common/netif.cpp:
```diff
@@ -82,42 +94,77 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
+        int64_t recv_result;
+        do {
+            recv_result = sock->Recv(temp, sizeof(temp), 0);
+        } while (recv_result < 0 && (errno == EINTR || errno == EAGAIN));
```
Since you're touching this line... According to the internet, we should also check `EWOULDBLOCK` even though it's usually the same as `EAGAIN`, and it's likely not relevant for any system we support.
https://stackoverflow.com/a/49421517

Sjors reviewed (commented):
Although the code this PR touches isn't compiled on macOS, I did briefly check that things still work there. I also briefly tested on Ubuntu 24.10.

Left some inline question to wrap my head around the changes and refresh my memory of the original...

willcl-ark commented:
Thanks for the review @Sjors & @laanwj.

Going to move to draft while I re-work it a little.

willcl-ark committed:
net: handle multi-part netlink responses

Handle multi-part netlink responses to prevent truncated results from
large routing tables.

Previously, we only made a single recv call, which led to incomplete
results when the kernel split the message into multiple responses (which
happens frequently with NLM_F_DUMP).

Also guard against a potential hanging issue where the code would
indefinitely wait for NLMSG_DONE for non-multi-part responses by
detecting the NLM_F_MULTI flag and only continue waiting when necessary.

willcl-ark committed:
net: filter for default routes in netlink responses

Filter netlink responses to only consider default routes by checking the
destination prefix length (rtm_dst_len == 0).

Previously, we selected the first route with an RTA_GATEWAY attribute,
which for IPv6 often resulted in choosing a non-default route instead of
the actual default.

This caused occasional PCP port mapping failures because a gateway for a
non-default route was selected.

willcl-ark committed:
net: skip non-route netlink responses

This shouldn't usually be hit, but is a good belt-and-braces.

willcl-ark commented:
After (additional) further investigation I gained some new insights:

- This is _not_ a regression from migration from `libnatpmp` as I first thought.
- The extent to which `tailscale` interferes with the routing is more significant than I realised:
  Even if we find the correct  gateway (by handling multi-part messages and filtering properly), the _route_ to the gateway still doesn't return the data we want, specifically the correct interface.
  This causes us to make an invalid request, which gets rejected.

  This is also observed using the `ip` command:
  ```bash
  $ ip route show
  default via 10.0.0.1 dev wlo1 proto dhcp src 10.0.12.100 metric 600
  10.0.0.0/20 dev wlo1 proto kernel scope link src 10.0.12.100 metric 600
  172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1
  192.168.122.0/24 dev virbr0 proto kernel scope link src 192.168.122.1 linkdown

  will@ubuntu in …/src/core/bitcoin on  pcp-default-multipart [$?⇕] : C v19.1.7-clang via △ v3.31.6 : 🐍 (core)
  $ ip route get 10.0.0.1
  10.0.0.1 dev tailscale0 table 52 src 100.89.20.73 uid 1000
      cache
  ```

  Handling this would be quite invasive, require tracking the interface, and I don't think it's in scope for us.

Therefore I have re-worked and split the remaining changes and into 3 commits:

1. Filter for the default route in the response. With this change ipv6 pinholing works even with `tailscale up`, which is quite nice.
2. Skip non `NEWROUTE` messages. Although `NLMSG_DONE` is already handled, this is a slight optimisation in protecting against unexpected responses being parsed.
3. Handle single and multi-part messages. Track if `NLM_F_MULTI` was set and if so wait for `NLMSG_DONE`, otherwise break after the first response, assuming a single-part response.

I think these are all worthwhile, but I could seean argument that what we currently have works "well enough" for the most basic use-case; a simple (single) routing table. And we could just consider anything else out-of-scope in this project.

@Sjors: Ref your comments, the socket is set as `nonblocking` already via the implementation in `CreateSockOS` in https://github.com/bitcoin/bitcoin/blob/74d9598bfbc24c3b7bfe1dad5bf9d988381bf893/src/netbase.cpp#L536-L540 I did [try a commit](https://github.com/bitcoin/bitcoin/commit/2f0b8d61b63edcaa1a9ac06b7e02cd3ea6e1e554) to add a backoff timer and retry mechanism to this for extra safety, but pretty sure it's not worth the added complexity.

fanquake commented:
@laanwj you might be interested in circling back here to review?

achow101 commented on src/common/netif.cpp:
```diff
@@ -84,41 +87,78 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
-                return CNetAddr(gw, scope_id);
+        bool processed_one{false};
+        for (nlmsghdr* hdr = (nlmsghdr*)response; NLMSG_OK(hdr, recv_result); hdr = NLMSG_NEXT(hdr, recv_result)) {
+            rtmsg* r = (rtmsg*)NLMSG_DATA(hdr);
+            int remaining_len = RTM_PAYLOAD(hdr);
```
These could be moved after `if (hdr->nlmsg_type == NLMSG_DONE) {` to avoid the integer sanitizer warning in #33245

davidgumberg commented on src/common/netif.cpp:
```diff
@@ -97,6 +97,11 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
+        // Only consider default routes (destination prefix length of 0).
+        if (r->rtm_dst_len != 0) {
+            continue;
+        }
```
Relevant (editorialized) code from iproute2:

```c
	if (tb[RTA_DST]) {
        // [...if the route has a specific destination, e.g.] ip route add 192.168.0.5 via  192.168.0.4
	} else if (r->rtm_dst_len) {
        // [if the route has prefix has a length]
		snprintf(b1, sizeof(b1), "0/%d ", r->rtm_dst_len);
	} else {
        // [if rtm_dst_len == 0]
		strncpy(b1, "default", sizeof(b1));
	}

https://github.com/iproute2/iproute2/blob/3337e5013e168bcf4fab5f6518d1e4293a0a830b/ip/iproute.c#L833-L851

davidgumberg commented on src/common/netif.cpp:
```diff
@@ -84,41 +87,78 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
+        do {
+            recv_result = sock->Recv(response, sizeof(response), 0);
+        } while (recv_result < 0 && (errno == EINTR || errno == EAGAIN));
+        if (recv_result < 0) {
```
nit: should also error on `recv_result == 0`, e.g. from libnetlink: 

```c
static int __rtnl_recvmsg(int fd, struct msghdr *msg, int flags)
{
	int len;

	do {
		len = recvmsg(fd, msg, flags);
	} while (len < 0 && (errno == EINTR || errno == EAGAIN));

	if (len < 0) {
		fprintf(stderr, "netlink receive error %s (%d)\n",
			strerror(errno), errno);
		return -errno;
	}

	if (len == 0) {
		fprintf(stderr, "EOF on netlink\n");
		return -ENODATA;
	}

	return len;
}
```
https://github.com/iproute2/iproute2/blob/3337e5013e168bcf4fab5f6518d1e4293a0a830b/lib/libnetlink.c#L813-L816:

davidgumberg commented on src/common/netif.cpp:
```diff
@@ -84,50 +87,78 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
+            for (rtattr* attr = RTM_RTA(r); RTA_OK(attr, remaining_len); attr = RTA_NEXT(attr, remaining_len)) {
+                if (attr->rta_type == RTA_GATEWAY) {
+                    rta_gateway = attr;
+                } else if (attr->rta_type == RTA_OIF && sizeof(int) == RTA_PAYLOAD(attr)) {
```
For context on this, IPv6 allows "scoped addresses", e.g. I can send a packet to a non-globally unique address, like a link-local address that has two destinations on two different links, so in order to disambiguate scoped addresses  [RFC4007#section-6](https://datatracker.ietf.org/doc/html/rfc4007#section-6) recommends the use of unique zone indexes:

>  Because the same non-global address may be in use in more than one zone of the same scope (e.g., the use of link-local address fe80::1 in two separate physical links) and a node may have interfaces attached to different zones of the same scope (e.g., a router normally has multiple interfaces attached to different links), a node requires an internal means to identify to which zone a non-global address belongs.  This is accomplished by assigning, within the node, a distinct "zone index" to each zone of the same scope to which that node is attached, and by allowing all internal uses of an address to be qualified by a zone index.

And  [RFC4007#section-7](https://datatracker.ietf.org/doc/html/rfc4007#section-7) suggests that specifying an output interface index as the zone index is sufficient, but more specific than necessary to disambiguate: 
>  When an upper-layer protocol sends a packet to a non-global destination address, it must have a means of identifying the intended zone to the IPv6 layer for cases in which the node is attached to more than one zone of the destination address's scope.
> Although identification of an outgoing interface is sufficient to identify an intended zone (because each interface is attached to no more than one zone of each scope), in many cases that is more specific than desired.

The name used in the rtnetlink interface (`RTA_OIF`) suggest that the unique scope ID / zone ID used  is the output interface index (OIF), but I guess that's an implementation detail, the point is, it's a unique index that makes an IPv6 address globally unique from the POV of a given device.

davidgumberg commented on src/common/netif.cpp:
```diff
@@ -84,50 +87,78 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
+            processed_one = true;
+            if (hdr->nlmsg_flags & NLM_F_MULTI) {
+                multi_part = true;
             }
```
nit:

`processed_one`, `multi_part` and the check at the bottom of the outer for loop could be dropped, with the following check:

```suggestion
            if (!(hdr->nlmsg_flags & NLM_F_MULTI)) {
                done = true;
            }
```

Sjors replied:
A slightly different approach, which avoids confusing `NLMSG_DONE` with `done`:

```diff
diff --git a/src/common/netif.cpp b/src/common/netif.cpp
index 77246980c3..3f3b8fdcea 100644
--- a/src/common/netif.cpp
+++ b/src/common/netif.cpp
@@ -117,9 +117,10 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
     // Receive response.
     char response[4096];
     ssize_t total_bytes_read{0};
-    bool done{false};
-    bool multi_part{false};
-    while (!done) {
+    // If we processed at least one message and multi flag not set, or if
+    // we received no valid messages, then we're done.
+    bool maybe_more{true};
+    while (maybe_more) {
         int64_t recv_result;
         do {
             recv_result = sock->Recv(response, sizeof(response), 0);
@@ -135,18 +136,16 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
             return std::nullopt;
         }
 
-        bool processed_one{false};
         for (nlmsghdr* hdr = (nlmsghdr*)response; NLMSG_OK(hdr, recv_result); hdr = NLMSG_NEXT(hdr, recv_result)) {
             rtmsg* r = (rtmsg*)NLMSG_DATA(hdr);
             int remaining_len = RTM_PAYLOAD(hdr);
 
-            processed_one = true;
-            if (hdr->nlmsg_flags & NLM_F_MULTI) {
-                multi_part = true;
-            }
+            if (!(hdr->nlmsg_flags & NLM_F_MULTI)) {
+                maybe_more = false;
+            };
 
             if (hdr->nlmsg_type == NLMSG_DONE) {
-                done = true;
+                maybe_more = false;
                 break;
             }
 
@@ -183,12 +182,6 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
                 }
             }
         }
-
-        // If we processed at least one message and multi flag not set, or if
-        // we received no valid messages, then we're done.
-        if ((processed_one && !multi_part) || !processed_one) {
-            done = true;
-        }
     }
 
     return std::nullopt;
```

davidgumberg commented on src/common/netif.cpp:
```diff
@@ -84,41 +87,78 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
-                std::memcpy(&scope_id, RTA_DATA(attr), sizeof(scope_id));
-            }
+        total_bytes_read += recv_result;
+        if (total_bytes_read > NETLINK_MAX_RESPONSE_SIZE) {
```
Feel free to mark resolved, this suggestion is actually incorrect, since when `recv_result == 0` the `NLMSG_OK` check [below](https://github.com/bitcoin/bitcoin/blob/4c531782569b42fc47478a468f4a79e0e2d93946/src/common/netif.cpp#L110) would fail:


```c
#define NLMSG_OK(nlh,len) ((len) >= (int)sizeof(struct nlmsghdr) && \
			   (nlh)->nlmsg_len >= sizeof(struct nlmsghdr) && \
			   (nlh)->nlmsg_len <= (len))
```
https://github.com/torvalds/linux/blob/39f90c1967215375f7d87b81d14b0f3ed6b40c29/include/uapi/linux/netlink.h#L107-L109

hodlinator commented:
Tested but did not dig deep into the code for now. Spent more than a day adding debug code in Bitcoin Core and fighting with my OpenWrt router (https://github.com/openwrt/packages/issues/17871#issuecomment-3243222244). Now that the router has been fully reset without any lingering configs from prior major versions, PCP & NATPMP is actually working.

On master (in 7/7 attempts roughly):
```
₿ cmake --build build -j16 --target=bitcoind && ./build/bin/bitcoind -regtest -debug=net
...
Bitcoin Core version v29.99.0-74d9598bfbc2 (release build)
[warning] Option '-upnp=true' is given but UPnP support was dropped in version 29.0. Substituting '-natpmp=true'.
...
[net] portmap: gateway [IPv4]: 192.168.1.1
[net] pcp: Requesting port mapping for addr 0.0.0.0 port 18444 from gateway 192.168.1.1
[net] pcp: Internal address after connect: 192.168.1.201
AddLocal([1111:1111::1]:18444,1)
Discover: 1111:1111::1
Bound to 127.0.0.1:18445
Bound to [::]:18444
Bound to 0.0.0.0:18444
Loaded 0 addresses from "anchors.dat"
0 block-relay-only anchors will be tried for connections.
init message: Starting network threads…
net thread start
init message: Done loading
opencon thread start
addcon thread start
msghand thread start
dnsseed thread start
Loading addresses from DNS seed dummySeed.invalid.
0 addresses found from DNS seeds
dnsseed thread exit
[net] pcp: Received response of 60 bytes: REDACTED...
[net:info] portmap: Added mapping pcp:< REDACTED  >:18450 -> 192.168.1.201:18444 (for 2400s)
AddLocal(< REDACTED  >:18450,3)
[net] portmap: gateway [IPv6]: fe80::REDACTED...
[net] pcp: Requesting port mapping for addr 1111:1111::1 port 18444 from gateway fe80::REDACTED...
[net] pcp: Internal address after connect: 1111:1111::1
[net] pcp: Received response of 60 bytes: REDACTED...
[net:warning] pcp: Mapping failed with result ADDRESS_MISMATCH (code 12)
```
On the router side:
```
... daemon.err miniupnpd[14402]: PCP: Invalid PCP v2 MAP message.
```
Taking the first commit (57ce645f05d18d8ad10711c347a5989076f1f788) from this PR makes the output towards the end better (my ISP doesn't support IPv6):
```
[net] pcp: Received response of 60 bytes: REDACTED...
[net:info] portmap: Added mapping pcp:< REDACTED  >:18464 -> 192.168.1.201:18444 (for 2400s)
AddLocal(< REDACTED  >:18464,3)
[net] portmap: Could not determine IPv6 default gateway
```
And there is no longer any error in the router log.

So PCP+IPv4 mostly works regardless of the PR, but at least the failure path for IPv6 is improved by it.

Haven't really noticed any changes in behavior with my setup from the latter 2 commits.

Might focus on other things for now as this already has 3 ACKs.

Sjors commented:
> Now that the router has been fully reset without any lingering configs from prior major versions, PCP & NATPMP is actually working.

You would not be the first person running into router bugs while testing NAT punching functionality :-)

davidgumberg commented:
Code Review re-ACK 88db09b

```console
$ git range-diff 4c53178...88db09b
```

https://github.com/bitcoin/bitcoin/commit/88db09bafe9ec363525e5e526c5f6cdd13691447 takes the [reviewer suggestion](https://github.com/bitcoin/bitcoin/pull/32159#discussion_r2302222710) of moving the `rtmsg*` cast after checking if `hdr->nlmsg_type == NLMSG_DONE`which resolves https://github.com/bitcoin/bitcoin/issues/33245 and implements [a suggested refactor](https://github.com/bitcoin/bitcoin/pull/32159#discussion_r2308655496) that simplifies the multi-message/done state tracking in the `NLMSG_NEXT` loop.
